### PR TITLE
SciPy integrate module that can be run on CUDA

### PR DESCRIPTION
**Pull request by Morteza-Khazaei**
Dear SciPy developer team,

I hope this message finds you well. I am writing to inquire about the possibility of adding a new option to Fortran QUADPACK in the SciPy integrate module that can be run on CUDA.

As you may know, CUDA is a parallel computing platform and programming model developed by NVIDIA for general computing on graphical processing units (GPUs). With the increasing availability of CUDA-enabled GPUs, there is growing interest in using them to accelerate scientific computations.

I believe that adding a CUDA-enabled option to the Fortran QUADPACK algorithm in SciPy integrate would be a valuable addition, allowing users to take advantage of the high parallelism offered by CUDA-enabled GPUs. This could greatly speed up the integration of functions, especially for large-scale problems.

I would be interested to know if this is something that the SciPy developer team is currently working on or considering for future development. If so, I would be happy to contribute to this effort in any way I can.

Thank you for your time and consideration.

Best regards,
Morteza

### CODE REVIEW

**Comment by tupui** [comment]
Hi @Morteza-Khazaei, thank you for reporting.

I think you made a mistake with your branch. The diff is huge and also the PR should be against `main`. I will close this now because we cannot change the target branch on an opened PR.

As it is I am not sure what this PR would propose.

Before opening another PR, I would instead suggest that you send an email to our mailing list. Such proposal would need first to be discussed before anything.

**Comment by Morteza-Khazaei** [comment]
Dear SciPy developer team,

Thank you for your response. I apologize for any confusion caused by my
previous message. I understand that my previous message may have been
unclear or may have given the impression that I had submitted a pull
request related to adding a CUDA option to the Fortran QUADPACK algorithm.

To clarify, I have not submitted a pull request related to this matter.
Rather, I was inquiring about the possibility of adding a CUDA option to
the Fortran QUADPACK algorithm in SciPy integrate.

I appreciate your suggestion to send an email to the mailing list to
discuss this proposal. I will do so to initiate a discussion regarding the
possibility of adding a CUDA option to the Fortran QUADPACK algorithm in
SciPy integrate.

Thank you for your time and consideration, and I look forward to further
discussion on this matter.

Best regards,

Morteza

**Comment by tupui** [comment]
I am not an expert but what I know on the topic is that GPU specific code would instead go into CuPy. We have a mechanism in place to have some interoperability between the two projects.

**Comment by mdhaber** [comment]
> GPU specific code would instead go into CuPy

Hmm my understanding was that we were investigating the possibility of having some SciPy functions work with CuPy arrays as part of the NASA work. In many cases (e.g. `scipy.stats.bootstrap`), this is almost as simple as calling the CuPy equivalents of NumPy functions/methods, and it could be done without having to maintain two separate versions of the code (e.g. use `xp` instead of `np` throughout, and `xp` refers to `cupy` when the array is a CuPy array). I'm not sure about QUADPACK in particular, but there is a lot in integrate (and stats, optimize) that I've wanted to adapt like this. Did I misunderstand @rgommers?

**Comment by rgommers** [comment]
That sounds about right. CuPy has a goal of implementing all SciPy APIs for GPUs (CUDA & ROCm at the moment). There's an open issue for `integrate`: https://github.com/cupy/cupy/issues/7019. And that is also linked from the overall SciPy APIs tracker.

**Comment by Morteza-Khazaei** [comment]
Sounds good.
But, I was about to find a way to optimize quadpack source with ability to
be run on the gpu, and I found out that it need huge modifications in the
Fortran code! Therefore, I ask this question to be aware of your plans
about integrate part of scipy. Also, I saw the cupy optimization for numpy
and scipy but it wasn't considering numerical integration of scipy yet!

**Comment by mdhaber** [comment]
> That sounds about right.

Just to confirm, you mean what @tupui wrote is right (not what I wrote)?

**Comment by ev-br** [comment]
If quad-like functionality is indeed desired on GPU (which I suspect needs measuring, is there that much room for GPU parallelism really?), I'd suggest as a way forward to port `quad_vec` to CuPy. It's pure python code, so it should be way easier, most likely it's just replacing `np.*` with `cupy.*` analogs. If CuPy  has an open issue than it's likely they'll accept it.

Porting QUADPACK Fortran sources to CUDA Fortran is a rather more difficult task. Even if it's done, SciPy is very unlikely to accept it. Whether CuPy accepts Cuda Fortran I don't know but would not be suprised if not.

**Comment by mdhaber** [comment]
> If quad-like functionality is indeed desired on GPU (which I suspect needs measuring, is there that much room for GPU parallelism really?),

Maybe not `quad`, but for something like `qmc_quad` and the ODE solvers (especially explicit ones) I think so. I've even done [work](https://zenodo.org/record/3961198#.ZANDs3bMJD8) in which I got huge speedups evaluating only the integrand on the GPU, passing data back and forth each time. Certainly it would be even better if SciPy could leave it on the GPU.

I support CuPy creating GPU versions of SciPy functions in SciPy code can't easily be dual-purpose (e.g. scipy.special and anything that requires writing a kernel for the GPU), but I'd prefer to see us be flexible about the underlying array library where possible.

**Comment by ev-br** [comment]
Let me ping @rgommers for our previous off-line discussions about trial porting things to `array_api`.

**Comment by Morteza-Khazaei** [comment]
Dear Scipy dev team,

I am writing to inquire about the potential benefits of using Cupy to move
the computation part of the input function of dblquad to the GPU. I have
read on your website that using low-level callback functions can
significantly improve integration speed, and I am wondering if using Cupy
can provide even further improvements.

I have some experience with Cupy and have noticed significant speedups in
computations involving large arrays. Given that the input function of
dblquad may involve complex computations, I am curious if using Cupy to
move the computation to the GPU can help reduce the overall computation
time.

I would greatly appreciate your thoughts on this matter and any guidance
you can provide on how to approach this. Thank you for your time and for
all the amazing work you do on Scipy.

Best regards,
Morteza

**Comment by rkern** [comment]
I'm afraid that this did not send an email to the `scipy-dev` mailing list. Rather, it continued to add to the [PR that you apparently accidentally created](https://github.com/scipy/scipy/pull/18091). Please go here to subscribe to `scipy-dev` and post your question:

  https://mail.python.org/mailman3/lists/scipy-dev.python.org/

**Comment by rgommers** [comment]
> Just to confirm, you mean what @tupui wrote is right (CuPy will create a copy of all SciPy within CuPy) and not what I wrote (SciPy will make minimal modifications where possible to support both NumPy and other arrays with NumPy-compatible APIs)?

I meant "what both of you wrote". One does not exclude the other. CuPy does have a goal of implementing all SciPy APIs (at least the ones that are in decent shape / not deprecated / etc.) in `cupy.cupyx.scipy`. And also, we'd like to be able to accept CuPy (and PyTorch, and whatever is array API compliant) array input where possible - without carrying code specific to GPUs or to other array/tensor libraries.


> Let me ping @rgommers for our previous off-line discussions about trial porting things to `array_api`.

We should be (re)starting this effort soon. This is made quite a bit easier now by having https://github.com/data-apis/array-api-compat around which at the moment supports NumPy, CuPy and PyTorch. We'll wait for a little bit to see how incorporating that into scikit-learn goes (scikit-learn is already further along with array API adoption, see https://scikit-learn.org/stable/modules/array_api.html).

**Comment by tupui** [comment]
Ultimately why do we want to not include GPU specific code into SciPy? You could make a similar argument with compiled code saying that other libraries could take care of that instead of us. We could have everything in pure Python and rely on the array API for other libraries to provide accelerator. So why the cut on CPU vs GPU?

**Comment by rgommers** [comment]
@tupui there's at least three main reasons:
- Maintainability: reviewing and maintaining CUDA (and ROCm, and possibly other things like oneAPI for Intel GPUs etc. in the future) code is hard. Very few SciPy maintainers have any experience with this, and it's nontrivial to learn.
- CI availability: there are no free CI offerings.
- Scope management: SciPy's scope is single-node CPUs. No GPU or distributed code. We do have parallel code, but that's still fairly limited (e.g., no OpenMP, no defaulting to parallel execution). Changing the scope at that level probably doesn't make sense at this point in time - it's too heavy a lift, and there are already other projects focusing on this.

**Comment by tupui** [comment]
Did we have a public discussion around the topic? "GPU" does not return much when searching the mailing list.

The space is rapidly evolving and if external libraries like CuPy, Bodo, etc. duplicate our API good enough with some added benefits like GPU, Arrow, other?, this is going to be more and more difficult for SciPy to exist. Newcomers are already not very familiar with SciPy (my experience talking with newcomers at conf and coworkers.)

**Comment by rgommers** [comment]
> Did we have a public discussion around the topic?

I'm sure we touched on it multiple times, e.g. when adding the roadmap item on GPU and distributed array library support. No explicit discussion on including CUDA code directly perhaps, because it seems quite clear that that's not a healthy idea. I cannot remember any maintainer proposing this.

> The space is rapidly evolving and if external libraries like CuPy, Bodo, etc. duplicate our API good enough with some added benefits like GPU, Arrow, other?, this is going to be more and more difficult for SciPy to exist.

This is not a very realistic take. CuPy is GPU-only and very much complementary to SciPy. Bodo has no support for SciPy, and if they gain it then it'll also be complementary since it's a JIT compiler for existing libraries. JAX and PyTorch have some overlap, but only a fraction of the total SciPy API is covered.

I'm really not worried about SciPy having difficulties existing. And in (the unlikely) case another much faster and relatively complete library emerges, then that's probably a good thing for the whole ecosystem.
